lib/icingadb/icingadb-objects.cpp

```diff
@@ -2973,3 +3341,137 @@ void IcingaDB::DeleteRelationship(const String& id, const String& redisKeyWithou
 
 	m_Rcon->FireAndForgetQueries(queries, Prio::Config);
 }
+
+void IcingaDB::DeleteRelationship(const String& id, RedisKey redisKey, bool hasChecksum)
+{
+	switch (redisKey) {
+		case RedisKey::RedundancyGroup:
+			DeleteRelationship(id, "redundancygroup", hasChecksum);
+			break;
+		case RedisKey::DependencyNode:
+			DeleteRelationship(id, "dependency:node", hasChecksum);
+			break;
+		case RedisKey::DependencyEdge:
+			DeleteRelationship(id, "dependency:edge", hasChecksum);
+			break;
+		default:
+			BOOST_THROW_EXCEPTION(std::invalid_argument("Invalid RedisKey provided"));
+	}
+}
+
+void IcingaDB::DeleteState(const String& id, RedisKey redisKey, bool hasChecksum) const
+{
+	String redisKeyWithoutPrefix;
+	switch (redisKey) {
+		case RedisKey::RedundancyGroupState:
+			redisKeyWithoutPrefix = "redundancygroup:state";
+			break;
+		case RedisKey::DependencyEdgeState:
+			redisKeyWithoutPrefix = "dependency:edge:state";
+			break;
+		default:
+			BOOST_THROW_EXCEPTION(std::invalid_argument("Invalid state RedisKey provided"));
+	}
+
+	Log(LogNotice, "IcingaDB")
+		<< "Deleting state " << std::quoted(redisKeyWithoutPrefix.CStr()) << " -> " << std::quoted(id.CStr());
+
+	RedisConnection::Queries hdels;
+	if (hasChecksum) {
+		hdels.emplace_back(RedisConnection::Query{"HDEL", m_PrefixConfigCheckSum + redisKeyWithoutPrefix, id});
+	}
+	hdels.emplace_back(RedisConnection::Query{"HDEL", m_PrefixConfigObject + redisKeyWithoutPrefix, id});
+
+	m_Rcon->FireAndForgetQueries(std::move(hdels), Prio::RuntimeStateSync);
+	m_Rcon->FireAndForgetQueries({{
+		"XADD", "icinga:runtime:state", "MAXLEN", "~", "1000000", "*",
+		"redis_key", m_PrefixConfigObject + redisKeyWithoutPrefix, "id", id, "runtime_type", "delete"
+	}}, Prio::RuntimeStateStream, {0, 1});
+}
```

Comment:
As discussed in the Icinga DB PR (https://github.com/Icinga/icingadb/pull/889#discussion_r2007390541), I would avoid sending runtime state deletions until Icinga DB can actually handle them properly (which would be the same as the current behavior for host/service states).

With the current handling in Icinga DB, runtime state updates are performed in parallel which can result in insertions and deletions being executed out of order, which can result superfluous and missing rows at random. That underlying problem won't be solved by not sending runtime updates, but this pushes the behavior towards the safer variant where there can only be superfluous rows. This behavior is better because those rows simply aren't used while missing rows can result in information missing in Icinga Web.

Reply:
Should I just comment out the last `m_Rcon->FireAndForgetQueries()` call of that method or do you want me to remove it completely?

Reply:
> Should I just comment out the last `m_Rcon->FireAndForgetQueries()` call of that method

That combined with a short comment why and a reference to https://github.com/Icinga/icingadb/pull/894 should do the trick.